src/preamble.js

```diff
@@ -1669,6 +1669,46 @@ function lookupSymbol(ptr) { // for a pointer, print out all symbols that resolv
 var memoryInitializer = null;
 
 #if USE_PTHREADS
+
+// JS library / EM_ASM proxying
+
+function proxyToMainThread() {
+  var args = Array.prototype.slice.call(arguments);
+  // Arguments are:
+  //   * the index of the function
+  //   * the call args
+  //   * whether it is sync
+  var sync = args[args.length - 1];
+  // The serialization buffer contains the number of call params, and then
+  // all the args here.
+  // We also pass 'sync' to C separately, since C needs to look at it.
+  // The buffer remains alive until receiveOnMainThread frees it.
+  var bufferLen = args.length + 1;
+  var buffer = _malloc(bufferLen * 8); // TODO: stackAlloc if sync?
```

Comment:
What happens if `malloc` needs to grow memory? I guess `stackAlloc` would work for sync, maybe that's enough.

Reply:
I don't think we need any special handling for that here? We already do a malloc for such things in the C proxying code. What risk are you thinking of?

(One risk here could be holding on to a view before the malloc and using it afterwards, but we avoid it.)

Reply:
A `malloc()` per each proxied call can be catastrophic. We should not need to allocate memory for the args array, the `em_queued_call` object already contains memory for that. Instead this should be able to `stackAlloc()` both in sync and async cases, and in `emscripten_run_in_main_runtime_thread_js()` function, copy the args into the fields of the `em_queued_call` object to be allocationless?

Reply:
We do already do a malloc in several proxying code paths for each proxy operation (like `emscripten_async_waitable_run_in_main_runtime_thread_` etc, all of which do `em_queued_call_malloc`). I agree it may be worth looking to optimize all those, but perhaps it could be a later optimization, after we look at data to see if it's worth it?

Copying the args into the `em_queued_call` is an option, but it's a little hackish, I feel?

Reply:
`em_queued_call_malloc`+`em_queued_call_free` was created specifically as a custom allocation function so that it can be made to pool and not call to malloc on each allocation of the structure.

I'm not sure why you say reusing the args space in struct `em_queued_call` would be a hack? That space is specifically allocated there to be able to contain the args to the function.

Reply:
I suppose that's true. Ok, will rewrite to use the `em_queued_call` storage.